calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -383,6 +384,7 @@ func unmarshalResource(tm unstructured.Unstructured, b []byte) ([]runtime.Object
 		return nil, err
 	}
 
+	defaults.SetDefaults(unpacked)
```

Comment:
isn't this overriding the actually set values?

Reply:
From what I can tell it only overrides empty values.

```
	func (f *Filler) SetDefaultValues(fields []*FieldData) {
	for _, field := range fields {
		if f.isEmpty(field) {
			f.SetDefaultValue(field)
		}
	}
```

Reply:
What if the value was _explicitly_ set to the "empty value"? And the empty value is valid, but not the default…

I don't know if we have cases like that. :thinking:

Reply:
I'm not sure why we need this, we should be hitting this code in libcalico IIUC: https://github.com/projectcalico/libcalico-go/blob/master/lib/clientv3/ippool.go#L58

Reply:
Even if you do not have that usecase now, pulling in a lib that has such corner cases? If so, it should be used broadly. What if you have an int that by default should be 25 but 0 is a valid value? The annotation is cool, if only it was used by the yaml unmarshaller.

If used this way, should it go to before line 383? First create an uninitialized object, set default then unmarshall into this partially initialized struct?

Reply:
> I'm not sure why we need this, we should be hitting this code in libcalico IIUC: https://github.com/projectcalico/libcalico-go/blob/master/lib/clientv3/ippool.go#L58

@caseydavenport From my digging, it seems that calicoctl invoked apply via
https://github.com/projectcalico/calicoctl/blob/master/calicoctl/commands/resources.go#L129. `CreateResourcesFromFile` reads the file and unmarshals the content at https://github.com/projectcalico/calicoctl/blob/2140631e726895836474beb1e5ffcbbf6a0eb737/calicoctl/resourcemgr/resourcemgr.go#L457-L497.  Then the validation fails at https://github.com/projectcalico/calicoctl/blob/master/calicoctl/resourcemgr/resourcemgr.go#L407.

Reply:
If validation fails, because something is not set properly, that is correct, right? Or at least you can see it that way ...

Reply:
@tomastigera  Looking over the docs for ippool at https://github.com/projectcalico/calico/blob/60e262b977e5338242bb9b3c6670626364f79f5e/v3.7/reference/calicoctl/resources/ippool.md the default value should be `Never`. 

I applied your suggestions about setting the defaults before unmarshalling. Please take a look.